README.md
# kue-jm
[![Build Status](https://travis-ci.org/Wiredcraft/kue-jm.svg?branch=master)](https://travis-ci.org/Wiredcraft/kue-jm)

  a J(ob) M(anager) tool for handling tasks(sub job) sequentially in Kue.js
 

## Installation
 `npm i`
 
 `npm test`
 
## Feature
 - Base on Kue.js
 - Support redis sentinel
 - Allow parent-child task design
 - Auto revert tasks to clean side efforts

## Examples
   if you want see more detail information, please set `DEBUG=jm:*` ;
   

## TODO
 - [X] implemented with event emitter.
 - [ ] support more job status query.
 - [ ] support more error handler.

CHANGELOG.md
## Change Log

### v1.0.1 (2019/05/05 05:47 +00:00)
- [#16](https://github.com/Wiredcraft/kue-jm/pull/16) add redis auth password to constructor options (@woodpig07)

### v1.0.0 (2019/03/08 05:47 +00:00)
- [#14](https://github.com/Wiredcraft/kue-jm/pull/14) add ttl for each task key (@ChopperLee2011)
- [#12](https://github.com/Wiredcraft/kue-jm/pull/12) add setGracefulShutdown method (@ChopperLee2011)

test/config.js
module.exports = {
  type: 'normal',
  redisHost: '127.0.0.1',
  redisPort: 6379,
  redisDBForJob: 1,
  redisDBForTask: 0,
  debug: false,
};


test/fixture/task1.js
module.exports = (param, previousResult, callback) => {
  setTimeout(() => {
    callback(null, JSON.parse(param).foo);
  }, 50);
};


test/fixture/task2.js
module.exports = (param, previousResult, callback) => {
  setTimeout(() => {
    callback(null, previousResult + param.baz);
  }, 10);
};


lib/jobManager.js
'use strict';

const EventEmitter = require('events');
const util = require('./util');
const redisFactory = require('./redisFactory');
const TaskSeries = require('./taskSeries');
const debug = require('debug')('jm:lib:jobmanager');

class JobManager extends EventEmitter {

  constructor(opt) {
    super();
    opt = opt || {};
    this.job = { _db: redisFactory.job(opt) };
    this.task = { _queue: redisFactory.task(opt), collections: new Map() };
    this.debug = opt.debug || false;
    this.handlers = new Map();
  }

  /**
   * Add a job to the job manager.
   * @param {String} job type
   * @param {Object} job data
   * @param {Object} options
   * @return added job instance
   */
  addJob(type, data, tasks, opt) {
    debug('addJob type: %s \t data: %o \t tasks: %o \t opt: %o', type, data, tasks, opt);
    if (!type || typeof type !== 'string') {
      throw new Error('please pass correct param');
    }
    opt = opt || {};
    const uniqField = opt.uniqField || 'id';
    if (!data[uniqField]) {
      throw Error(`data[${uniqField}] can not be empty.`);
    }
    const key = `${type}:${uniqField}:${data[uniqField]}`;
    const jobTTL = opt.jobTTL || 1000 * 60 * 5;
    const resultTTL = opt.resultTTL || 3600 * 24 * 3;
    const jobDb = this.job._db;
    return jobDb.set(key, 'init')
      .then(() => {
        debug('addJob set key: %s', key);
        data.tasks = tasks;
        const task = this.task._queue
          .create(type, data)
          .removeOnComplete(!this.debug)
          .ttl(jobTTL);
        task
          .on('complete', (result) => {
            debug('addJob complete result: %o', result);
            this.emit('complete', result);
            if (typeof result === 'object') {
              result = JSON.stringify(result);
            }
            if (resultTTL < 0) {
              jobDb.set(key, result);
            } else {
              jobDb.setex(key, resultTTL, result);
            }
          })
          .on('failed', (errorMsg) => {
            debug('addJob job error: %s', errorMsg);
            this.emit('failed', errorMsg);
          })
          .save((err) => {
            if (err) {
              jobDb.del(key);
              return Promise.reject(new Error(`save job error, ${err.stack}`));
            }
            debug('addJob save job key: %s \t job id: %s', key, task.id);
            jobDb.set(key, task.id);
            // TODO: better not have this set op.
            debug('addJob resolve job.id: %o', task.id);
            return Promise.resolve(true);
          });
      });
  }

  /**
   * Get status of tasks relate to a specific job
   * If those taks completed, an array of empty objects will be resolved
   * If one of the tasks failed, an array of task infomatiexecute
   * @param {string} uniqFiledValue
   * @return {promise} a promise resolved to an array
   */
  getTaskStatus(jobType, uniqFieldValue) {
    const num = this.listTasks(jobType).length;
    const statusPromises = [];
    for (let i = 0; i < num; i++) {
      const status = this.job._db
        .hgetall(`${jobType}:${uniqFieldValue}:${i}`);
      statusPromises.push(status);
    }
    return Promise.all(statusPromises);
  }

  /**
   * List tasks for the job type.
   * @param {string} job type.
   * @return
   */

  listTasks(jobType) {
    debug('getTasks jobType: %s', jobType);
    if (!jobType || typeof jobType !== 'string') {
      throw new Error('please pass correct param');
    }
    return this.job._db.get(`jobs:${jobType}`);
  }

  /**
   * Remove task for the job type.
   * @param {string} job type
   * @param {string} task name
   * @return
   */

  removeTask(jobType, taskName) {
    throw new Error(`please implement this for ${taskName}`);
  }

  /**
   * watchStuckJobs.
   * @param {Number} interval
   * @return
   */
  watchStuckJobs(interval) {
    interval = interval || 1000;
    const orgPrefix = this.task._queue.client.prefix;
    this.task._queue.client.prefix = 'q';
    this.task._queue.watchStuckJobs(interval);
    this.task._queue.client.prefix = orgPrefix;
  }

  /**
   * Run one job.
   * @param {string} job type
   * @param {number} concurrency
   * @param {number} ttl for each job key in redis
   * @return {string} job uuid
   */
  run(type, concurrency = 1, taskTTL = 3600 * 24 * 180) {
    debug('run type: %s \t concurrency: %s', type, concurrency);
    return new Promise((resolve, reject) => {
      if (!type || typeof type !== 'string') {
        return reject(new Error('please pass correct param'));
      }
      this.task._queue.process(type, concurrency, (job, done) => {
        const series = new TaskSeries(this);
        const tasks = job.data.tasks;
        if (!tasks || tasks.length === 0) {
          done(null, null);
          return resolve();
        }
        const multi = this.job._db.multi();
        const sid = `${type}:${job.data.id}`;
        debug('run with sid: %o', sid);
        tasks.forEach((task, i) => {
          tasks[i].idx = i;
          multi.hmset(`${sid}:${i}`, util.serialize(task));
          multi.expire(`${sid}:${i}`, taskTTL);
        });
        multi.exec((err) => {
          if (err) {
            return reject(err);
          }
          // TODO: blocking queue processing have bad pref, should change it.
          series.execute(this, sid, taskTTL)
            .then((res) => {
              debug('execute res %o', res);
              this.cleanWorker(sid);
              done(null, res);
            })
            .catch((err) => {
              debug('execute err: %o', err);
              this.cleanWorker(sid);
              done(err);// main job
            });
          resolve(sid);
        });
      });
    });
  }

  cloneJob() {
    throw new Error('this API is not implement yet');
  }

  toJSON() {
    throw new Error('this API is not implement yet');
  }

  /**
   * clean finished worker.
   */
  cleanWorker(type) {
    debug('cleanWork type: %s', type);
    if (this.task._queue.workers) {
      const idx = this.task._queue.workers.findIndex(worker => worker && worker.type === type);
      if (idx !== -1) {
        this.task._queue.workers[idx].shutdown();
        this.task._queue.workers[idx] = undefined;
      }
    }
  }

  registerHandler(name, handler) {
    debug('registerHandler name: %s', name);
    if (this.handlers.get(name)) {
      throw new Error('this handler name is already token');
    }
    this.handlers.set(name, handler);
  }

  setGracefulShutdown(timeout = 5000) {
    process.once('SIGTERM', () => {
      this.tasks._queue.shutdown(timeout, (err) => {
        debug('Kue shutdown with err %j', err);
        process.exit(0);
      });
    });
  }
}

module.exports = JobManager;


lib/redisFactory.js
'use strict';

const Redis = require('ioredis');
const kue = require('kue');
const debug = require('debug')('jm:lib:redisFactory');

const getSentinels = (sentinelsString) => {
  debug('sentinelsString: %o', sentinelsString);
  return sentinelsString.split(',').map((each) => {
    return {
      host: each.split(':')[0],
      port: each.split(':')[1],
    };
  });
};

exports.job = (opt) => {
  debug('job opt: %o ', opt);
  let redis;
  if (opt.type === 'sentinel') {
    redis = new Redis({
      sentinels: getSentinels(opt.redisSentinelForJob),
      name: opt.redisSentinelNameForJob,
      password: opt.redisPassword
    });
  } else {
    redis = new Redis({
      password: opt.redisPassword
    });
  }
  redis.select(opt.redisDBForJob || 1);
  return redis;
};

exports.task = (opt) => {
  const queue = kue.createQueue({
    prefix: opt.prefix || 'q',
    redis: {
      createClientFactory: () => {
        if (opt.type === 'sentinel') {
          return new Redis({
            sentinels: getSentinels(opt.redisSentinelForJob),
            name: opt.redisSentinelNameForJob,
            db: opt.redisDBForTask || 0,
            password: opt.redisPassword
          });
        }
        return new Redis({
          db: opt.redisDBForTask || 0,
          password: opt.redisPassword
        });
      },
    },
  });
  queue.setMaxListeners(queue.getMaxListeners() + 1);
  return queue;
};


lib/taskSeries.js
'use strict';

const util = require('./util');
const debug = require('debug')('jm:lib:taskSeries');

class taskSeries {
  constructor(jm) {
    this.jm = jm;
  }

  /**
   * execute tasks
   * @param {jobManager} jm
   * @param {string} job unique id
   * @param {number} ttl for each task key in redis
   * @return
   */
  execute(jm, sid, taskTTL) {
    debug('execute tasks sid: %s \t', sid);
    jm.task._queue.process(sid, (job, done) => {
      debug('execute process job.data: %o', job.data);
      if (!job.data.result && job.data.status === 'complete' && job.data.result) {
        debug('execute get an complete job with data: %o', job.data.result);
        return done(null, job.data.result);
      }
      try {
        let mod;
        if (!job.data.rewindFlg) {
          mod = jm.handlers.get(job.data.handler);
        } else {
          mod = jm.handlers.get(job.data.rewindHandler);
        }
        const input = job.data.preResult;
        debug('execute task input: %o', input);
        // todo: throw error when param is empty;
        const param = job.data.param || {};
        mod(JSON.parse(param), input, (err, res) => {
          if (err) {
            debug('execute with error: %o', err);
            return done(err);
          }
          done(null, res);
        });
      } catch (e) {
        done(e);
      }
    });

    return this.createStep(jm, sid, 0)
      .catch((err) => {
        // TODO log err
        return this._processRewind(sid, err.index, taskTTL)
          .catch((e) => {
            debug('after rewind fail %s', e);
            throw e;
          })
          .then(() => {
            debug('after rewind sucess %s', err.message);
            throw err.message;
          });
      });
  }

  /**
   * create steps
   * @param {jobManager} jm
   * @param {task} task
   * @param {callback] cb
   * @return
   */

  createStep(jm, sid, idx, preResult) {
    debug('createStep sid: %s \t idx: %d \t preResult: %o', sid, idx, preResult);
    return new Promise((resolve, reject) => {
      jm.job._db.hgetall(`${sid}:${idx}`)
        .then((task) => {
          if (preResult) {
            task.preResult = preResult;
          }
          debug('createStep get task: %o', task);
          const job = jm.task._queue
            .create(sid, task)
            .removeOnComplete(!jm.debug)
            .attempts(task.retry)
            .ttl(task.ttl);
          job
            .on('complete', (result) => {
              debug('createStep complete idx: %s \t result: %o', idx, result);
              jm.job._db
                .multi([
                  ['hset', `${sid}:${idx}`, 'status', 'complete'],
                  ['hset', `${sid}:${idx}`, 'result', JSON.stringify(result)],
                ])
                .exec()
                .then(() => {
                  return this.next(jm, sid, idx);
                })
                .then((nextTask) => {
                  resolve({ result, nextTask });
                })
                .catch(err => reject(err));
            })
            .on('failed', (errMessage) => {
              debug('createStep job failed');
              jm.job._db
                .multi([
                  ['hset', `${sid}:${idx}`, 'status', 'failed'],
                  ['hset', `${sid}:${idx}`, 'error', errMessage],
                ])
                .exec();
              return reject({ message: errMessage, index: idx });
            })
            .save((err) => {
              if (err) {
                return reject(err);
              }
              debug('createStep job saved');
            });
        });
    })
      .then((res) => {
        debug('createStep nextTask %o \t result: %o', res.nextTask, res.result);
        if (res.nextTask.idx) {
          return this.createStep(jm, sid, +res.nextTask.idx, res.result); // cast idx to number
        }
        return res.result;
      });
  }

  /**
   * next step
   * @param {jobmanager} jm
   * @param {sid} job unique id
   * @param {curIdx} current step id
   * @return
   */
  next(jm, sid, curIdx) {
    debug('next sid: %s \t curIdx: %d', sid, curIdx);
    return new Promise((resolve, reject) => {
      jm.job._db.hgetall(`${sid}:${++curIdx}`)
        .then((task) => {
          debug('next task: %o', task);
          resolve(util.deserialize(task));
        })
        .catch(err => reject(err));
    });
  }

  /**
   *
   * @param sid
   * @param errIdx
   * @param {number} ttl for each job key in redis
   * @returns {Promise}
   * @private
   */
  _processRewind(sid, errIdx, taskTTL = 3600 * 24 * 180) {
    debug('_processRewind sid: %s \t errIdx: %d \t ', sid, errIdx);
    let curIdx = errIdx;
    // init rewind tasks
    const multi = this.jm.job._db.multi();
    const rewindTasks = [];
    for (let idx = errIdx; idx >= 0; idx--) {
      rewindTasks.push(
        this.jm.job._db.hgetall(`${sid}:${idx}`)
          .then((task) => {
            debug('_processRewind task: %o', task);
            if (task.rewindHandler) {
              task.rewindFlg = true;
              task.idx = ++curIdx;
              const record = multi.hmset(`${sid}:${curIdx}`, task);
              multi.expire(`${sid}:${curIdx}`, taskTTL);
              return record;
            }
          })
      );
    }

    return Promise.all(rewindTasks)
        .then((res) => {
          debug('_processRewind finished setup rewind tasks');
          const tasks = res.filter(r => !!r);
          if (tasks.length === 0) {
            debug('_processRewind no rewind task');
            return multi.exec();
          } else {
            // stop old task process
            multi.del(`${sid}:${curIdx + 1}`);
            return multi.exec()
              .then(() => {
                return errIdx > 0 && this.jm.job._db.hgetall(`${sid}:${errIdx - 1}`);
              })
              .then(task => task && task.result && JSON.parse(task.result))
              .then(input => this.createStep(this.jm, sid, errIdx + 1, input));
          }
        });
  }
}

module.exports = taskSeries;
